ca/ca.go

```diff
@@ -609,9 +629,34 @@ func (ca *certificateAuthorityImpl) issuePrecertificateInner(ctx context.Context
 		return nil, nil, berrors.InternalServerError("failed to sign precertificate: %s", err)
 	}
 
+	cert, err := x509.ParseCertificate(certDER)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to parse precertificate bytes: %s", err)
+	}
+
+	err = verifyTBSCertificateDeterminism(lintCert, cert)
+	if err != nil {
+		return nil, nil, err
+	}
+
 	ca.signatureCount.With(prometheus.Labels{"purpose": string(precertType), "issuer": issuer.Name()}).Inc()
 	ca.log.AuditInfof("Signing precert success: issuer=[%s] serial=[%s] regID=[%d] names=[%s] precertificate=[%s] certProfileName=[%s] certProfileHash=[%x]",
 		issuer.Name(), serialHex, issueReq.RegistrationID, strings.Join(csr.DNSNames, ", "), hex.EncodeToString(certDER), certProfile.name, certProfile.hash)
 
 	return certDER, &certProfileWithID{certProfile.name, certProfile.hash, nil}, nil
 }
+
+// verifyTBSCertificateDeterminism verifies that x509.CreateCertificate signing
+// operation is deterministic and produced identical DER bytes between the given
+// lintCert and leafCert. If this fails it's mississuance, but it's better to
+// know about the problem sooner than later.
+func verifyTBSCertificateDeterminism(lintCert *x509.Certificate, leafCert *x509.Certificate) error {
```

Comment:
In both callsites (and the tests!), neither argument to this function is used for any other purpose. So this could easily take both arguments as []byte and handle the parsing inside here. This would make the test setup simpler, and remove the need to have `cert` and `lintCert` variables floating around inside the already-complex `issuePrecertificateInner` and `IssueCertificateForPrecertificate`.
```go
func verifyTBSCertificateDeterminism(lintCert []byte, leafCert []byte) error {
```


If you _really_ wanted to, you could take this a step further: rather than parsing the full cert and then comparing the RawTBSCertificate field, you could use cryptobyte yourself to extract it. Something like

```go
lintStr := cryptobyte.String(lintCert)
if !lintStr.ReadASN1(&lintStr, cryptobyte_asn1.SEQUENCE) {  // Unwrap the outer Certificate
	return nil, errors.New("x509: malformed certificate")
}
var lintTBS cryptobyte.String
if !lintStr.ReadASN1Element(&lintTBS, cryptobyte_asn1.SEQUENCE) {  // Extract the TBSCertificate
	return nil, errors.New("x509: malformed tbs certificate")
}
// Do the same for the final cert
// ...
// Note that cryptobyte.String is just an alias for []byte, so using bytes.Equal here works.
if !bytes.Equal(lintTBS, leafTBS) { ... }
```
This would be more of our own code to maintain, and a little bit of code duplication, but with the benefit of likely being significantly faster than letting the x509 library parse the whole thing.

Reply:
I waffled between using `lintCert []byte, leafCert []byte` and the current implementation. I knew it was going to get questioned 😆 . I'll go the cryptobyte route because I am a sucker for that package.

Reply:
I'm accepting alternate names for this function too.